app/Http/Livewire/ModuleContent.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace App\Http\Livewire;
+
+use App\Models\Module;
+use App\Models\User;
+use Illuminate\Support\Facades\Auth;
+use Livewire\Component;
+
+class ModuleContent extends Component
+{
+    public Module $module;
+
+    public User $user;
+
+    public ?string $startedContentAt = null;
+
+    public ?string $finishedContentAt = null;
+
+    public function mount(Module $module)
+    {
+        $this->module = $module;
+        $this->user = Auth::user();
+        $moduleUser = $this->user->modules->find($module->id);
+        $this->startedContentAt = $moduleUser?->getRelationValue('pivot')->started_content_at;
+        $this->finishedContentAt = $moduleUser?->getRelationValue('pivot')->finished_content_at;
+    }
+
+    protected $listeners = ['onPlayerEnd', 'onPlayerStart'];
+
+    public function onPlayerEnd()
+    {
+        if (! $this->finishedContentAt) {
+            $this->user->modules()->updateExistingPivot(
+                $this->module->id, ['finished_content_at' => now()]
+            );
+            $this->finishedContentAt = 'finished_content';
+            $finishedCourse = true;
+            $course = $this->module->course;
+            foreach ($course->modules as $module) {
+                $moduleUser = $module->users->find($this->user->id)?->getRelationValue('pivot');
+                if (! $moduleUser) {
+                    $finishedCourse = false;
+                    break;
+                } elseif (! $moduleUser->finished_content_at) {
+                    $finishedCourse = false;
+                    break;
+                }
+            }
+            if ($finishedCourse) {
+                $course->users()->updateExistingPivot(
+                    $this->user->id, ['finished_at' => now()]
+                );
+            }
+        }
```

Comment:
Another thought is to not have this in the pivot at all, unless the `finished_at` time is actually important but even then it could be defined from the latest module finished at time. Also there is some ambiguity here because the course is marked as finished when the modules are complete but not when the quiz is complete. So maybe you just have a function that you fun to see if all the courses modules are complete for the user. This method could be used for the policy checks I mention elsewhere too. Another thing with the existing code is that if a module is added after the user has finished the existing modules, it will still be marked as complete.

Reply:
It's probably also worth investigating to see if you can find only the modules that don't have a 'finished_content_at' in the module_user pivot. That way you can just see if the count is 0 to know if course is complete.

Reply:
I agree that we don't need finished_at attribute for the course, we could have a method similar to isPublishable, where it iterate through modules in courses and check their finished status to decide if a course is finished or not.